README.md
# apoptosisQuantification

The package `apoptosisQuantification` quantifies apoptosis and necroptosis
using predefined markers in proteomics datasets.

## Installation 

To install the package, enter

```
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("tnaake/apoptosisQuantification")
```

## Methods to quantify apoptosis/necroptosis

The package contains three ways to quantify apoptosis/necroptosis:

1. by comparing the distribution of marker loadings and non-marker 
   loadings (from PCA),
2. by looking at the contribution of mitochrondrial proteins
   (this might be misleading when looking at a proteomics data set),
3. by running `decoupleR` and calculating a score for 
   apoptosis/necroptosis.



inst/NEWS
Changes in version 0.99.0 (2021-02-07) 
+ implement quantification method by calculating the proportion of 
  mitochondrial proteins
+ implement quantification method by comparing the distribution of loadings 
  of markers and non-markers
+ implement quantification method by decoupleR, where contamination scores 
  are calculated
+ add functions for visualization for all three approaches
+ add unit tests for all functions
+ add a vignette showcasing the functionality using the data set of Tanzer et al. 
  (2020)

tests/testthat.R
library(testthat)
library(apoptosisQuantification)
testthat::test_check("apoptosisQuantification")

tests/testthat/test_markers.R
## function readMarkers
test_that("readMarkers works.", {
    apoptosis <- readMarkers(type = "apoptosis", fc = 1, n = 1)
    necroptosis <- readMarkers(type = "necroptosis", fc = 1, n = 2)
    expect_is(apoptosis, "tbl")
    expect_equal(colnames(apoptosis), 
        c("protein", "fold_change", "significant_n"))
    expect_equal(dim(apoptosis), c(89, 3))
    expect_equal(apoptosis$protein[1:5], 
        c("ACSL4", "ADNP", "AP1G1", "ASPSCR1", "ASRGL1"))
    expect_equal(apoptosis$fold_change[1:5], 
        c(2.477711, 2.057616, 2.040779, 1.479240, 2.152555),
        tolerance = 1e-06)
    expect_equal(apoptosis$significant_n[1:5], 
        c(1, 1, 1, 1, 1))
    expect_is(necroptosis, "tbl")
    expect_equal(colnames(necroptosis), 
        c("protein", "fold_change", "significant_n"))
    expect_equal(dim(necroptosis), c(253, 3))
    expect_equal(necroptosis$protein[1:5], 
        c("ACIN1", "ACOT13", "ACSL4", "ACY1;ABHD14A-ACY1", "ADAR"))
    expect_equal(necroptosis$fold_change[1:5], 
        c(2.117764, -2.305560, 1.440479, -2.364775, 3.851311),
        tolerance = 1e-06)
    expect_equal(necroptosis$significant_n[1:5], 
        c(3, 2, 2, 2, 3))
    
    ## type
    expect_error(readMarkers(type = "foo", fc = 1, n = 2),
        "'arg' should be one of ")
    
    ## fc
    expect_error(readMarkers(type = "apoptosis", fc = NULL, n = 2),
        "Must subset rows with a valid subscript vector")
    expect_equal(dim(readMarkers(type = "apoptosis", fc = "foo", n = 2)),
        c(0, 3))
    expect_equal(dim(readMarkers(type = "apoptosis", fc = 3, n = 1)), c(20, 3))
    
    ## n
    expect_error(readMarkers(type = "apoptosis", fc = 1, n = NULL),
        "Must subset rows with a valid subscript vector")
    expect_equal(dim(readMarkers(type = "apoptosis", fc = 1, n = "foo")),
        c(0, 3))
    expect_equal(dim(readMarkers(type = "apoptosis", fc = 1, n = 2)), c(5, 3))
})



R/markers.R
#' @name readMarkers
#' 
#' @title Load a file with known markers linked to apoptosis or necroptosis
#' 
#' @description
#' The function \code{readMarkers} loads a data set with markers linked to
#' apoptosis or necroptosis. 
#' 
#' The returned \code{tibble} contains three columns:
#' \itemize{
#'     \item \code{protein}: name of the marker, 
#'     \item \code{change}: maximum/minimum fold change from the 
#'         time-course experiment (maximum absolute value is taken), 
#'     \item \code{significant_n}: number of significant time-points.
#' }
#' 
#' Classical necrotic markers include lactate deydrogenase (LDH),
#' high-mobility group B1 (HMGB1), myoglobin, enolase, and 14-3-3 proteins.
#' Commonly, a protein release is observed during apoptosis and necroptosis
#' for human myeloid/lymphoma cell lines and human primary macrophage cells as
#' observed time course experiments (cf. Tanzer et al., 2020).  
#' 
#' @details 
#' The function loads the Supplementary Table S1 from Tanzer et al. (2020).
#' 
#' In the data set of Tanzer et al. (2020), the following abbreviations are 
#' used:
#' \itemize{
#'     \item TNF: tumor necrosis factor (TNF-mediated apoptosis),
#'     \item SM: birinapant  (inhibition of cIAPs using small molecules called 
#'         Smac mimetics),
#'     \item IDN: caspase inhibitor IDUN-6556 (IDN-6556, used in induction of 
#'         necroptosis),
#'     \item TNF application: leads to production of cytokines (no induction of 
#'         apoptosis or necroptosis),
#'     \item TNF+SM application: inducton of apptosis,
#'     \item TNF+SM+IDN-6556 application: induction of necroptosis.
#' }
#' 
#' The time points 1 h, 3 h, 5 h, and 7 h are taken into account for defining
#' the markers. For apoptosis, the time points 9 h, 12.5 h and 15 h are 
#' removed prior to filtering the data set. 
#' 
#' @references 
#' Tanzer et al. (2020): Quantitative and Dynamic Catalogs of Proteins 
#' Released during Apoptotic and Necroptotic Cell Death. 
#' \emph{Cell Reports}, 30,  1260-1270.e5. 10.1016/j.celrep.2019.12.079.
#' 
#' @param type character, \code{"apoptosis"} or \code{"necroptosis"}
#' @param fc numeric(1), threshold for fold change (default 2), e.g. if set to
#' 2, the proteins are retained that have absolute fold changes greater or equal
#' to \code{fc}
#' @param n numeric, threshold for number of significant time points
#' (default 1), e.g. if set to 1, the proteins are retained that have at 
#' least 1 significant time point
#' 
#' @return tibble
#' 
#' @importFrom dplyr pull
#' @importFrom tibble tibble
#' 
#' @export
#' 
#' @examples 
#' ## "apoptosis"
#' readMarkers(type = "apoptosis", fc = 2, n = 1)
#' 
#' ## "nectroptosis"
#' readMarkers(type = "necroptosis", fc = 2, n = 2)
readMarkers <- function(type = c("apoptosis", "necroptosis"), fc = 2, n = 1) {
    
    type <- match.arg(type)
    
    ## load the data set that contains the fold changes of the time-courses
    ## sheet "TNF+SM vs co"
    if (type == "apoptosis")
        f <- "protein_markers/Tanzer2020_protein_markers_apoptosis.RDS"
    ## sheet "TNF+SM+IDN-6556 vs co"
    if (type == "necroptosis")
        f <- "protein_markers/Tanzer2020_protein_markers_necroptosis.RDS"
    f <- system.file(f, package = "apoptosisQuantification")
    tbl <- readRDS(f)
    tbl <- tbl[!grepl(x = colnames(tbl), pattern = "9h|12[.]5h|15h")]
    cols <- colnames(tbl)
    
    ## remove 9h, 12.5h, 15h
    cols <- cols[!grepl(x = cols, pattern = "9h|12[.]5h|15h")]

    ## find the columns that contain the information on the significance
    sign_inds <- grep(x = cols, pattern = "Significant.TNF[+]SM")

    ## determine the proteins that are significant in at least n time
    ## course events, depending if type is apoptosis or necroptosis

    sign_logical <- apply(t(tbl[, sign_inds]), 1, 
        function(x) ifelse(is.na(x), FALSE, TRUE))
    sign_n <- apply(sign_logical, 1, sum)

    ## look now into the fold changes and find the absolute highest fold change
    ## for the significant proteins, store these fold changes in a list (fc_max)
    fc_inds <- grep(x = cols, pattern = "Fold.change.[(]Log2[)]")
    fc_tbl <- tbl[, fc_inds]
    fc_max <- lapply(seq_len(nrow(tbl)), function(row_i) {
        fc_tbl_i <- fc_tbl[row_i, ][sign_logical[row_i, ]]
        if (length(fc_tbl_i) > 0) 
            fc_tbl_i[[which.max(abs(fc_tbl_i))]]
        else 
            NA
    })
    fc_max <- unlist(fc_max)

    ## create a logical vector that stores information about which proteins
    ## to take (the ones that have significant times point greater or equal
    ## to n and that have fc_max > 0.5)
    markers_logical <- sign_n >= n & (!is.na(fc_max) & abs(fc_max) >= fc)
    
    ## return the tibble, store inside the protein name, the max (absolute) 
    ## fold change, and the number of time points the fold change was
    ## significant
    tibble::tibble(protein = dplyr::pull(tbl[markers_logical, ], "Gene.names"),
        change = fc_max[markers_logical],
        significant_n = sign_n[markers_logical])
}


R/quantifyViaLoadings.R
#' @name getLoadingsOfMarkers
#' 
#' @title Get PCA loadings of marker proteins 
#' 
#' @description
#' The function \code{getLoadingsOfMarkers} retrieves the loadings of the
#' marker proteins (defined by \code{markers}) found in \code{prot}.
#' 
#' \code{getLoadingsOfMarkers} returns a \code{tibble} of the marker 
#' proteins found in \code{prot} containing the loadings for the principal
#' components.
#' 
#' @details 
#' In case there are ambiguous/multiple assignments of proteins (containing 
#' markers) to a loading value (MaxQuant separates these assignments by ";"), 
#' the proteins that are not in \code{markers$protein} are removed.
#' In case there are multiple assignments left of marker proteins, the
#' mean value of the loading values are returned. 
#'
#' @references 
#' Tanzer et al. (2020): Quantitative and Dynamic Catalogs of Proteins 
#' Released during Apoptotic and Necroptotic Cell Death. 
#' \emph{Cell Reports}, 30, 1260-1270.e5. 10.1016/j.celrep.2019.12.079.
#'
#' @param prot matrix, data.frame, or tibble
#' @param markers tibble
#'
#' @return tibble
#'
#' @importFrom MatrixQCvis transformAssay imputeAssay
#' @importFrom methods is
#' @importFrom stats prcomp
#' @importFrom SummarizedExperiment assay
#' @importFrom tibble rownames_to_column tibble
#'
#' @export
#'
#' @examples 
#' library(MatrixQCvis)
#' library(SummarizedExperiment)
#' 
#' ## "apoptosis"
#' markers <- readMarkers(type = "apoptosis")
#' f <- system.file("protein_datasets/tanzer2020.RDS", 
#'     package = "apoptosisQuantification")
#' se <- readRDS(f)
#' prot <- assay(se) |>
#'     imputeAssay(method = "MinDet")
#' getLoadingsOfMarkers(prot = prot, markers = markers)
#' 
#' ## "nectroptosis"
#' markers <- readMarkers(type = "necroptosis")
#' getLoadingsOfMarkers(prot = prot, markers = markers)
getLoadingsOfMarkers <- function(prot, markers) {
    
    if (!methods::is(markers, "tbl"))
        stop("'markers' is not a tibble")
    
    ## calculate the loadings
    pca <- prot |> 
        as.matrix() |>
        t() |>
        stats::prcomp(center = TRUE, scale = TRUE)
    pca_rotation <- pca$rotation
    
    ## expand markers, some of the features are separated by ";", 
    ## split the strings and replicate the entries
    markers_protein <- splitNames(markers$protein, na.rm = FALSE)#strsplit(markers$protein, split = ";")
    markers_protein_len <- unlist(lapply(markers_protein, length))
    markers <- tibble::tibble(protein = unlist(markers_protein),
        fold_change = rep(markers$fold_change, times = markers_protein_len),
        significant_n = rep(markers$significant_n, times = markers_protein_len))
    
    ## remove duplicated entries
    markers <- markers[!duplicated(markers$protein), ]

    ## look up the names from pca_rotation and calculate the contribution of 
    ## the protein:
    
    ## split the names of the proteins (prot) and remove NA values
    prot_names <- splitNames(protein_names = rownames(pca_rotation), na.rm = TRUE)
    
    ## check if the prot_names are in markers$protein and remove the proteins
    ## that are not in markers
    prot_names <- lapply(prot_names, 
        function(prot_i) prot_i[prot_i %in% markers$protein])
    
    ## get the coordinate in PCn (loadings coordinates)
    marker_vals <- lapply(prot_names, function(prot_names_i) {
        names_i <- prot_names_i[prot_names_i %in% rownames(pca_rotation)]
        if (length(names_i) > 0)
            pca_rotation[names_i, ]
    })
    
    ## take the mean of the product (in case there are several, 
    ## indistinguishable proteins) or return the values as they are
    marker_vals <- lapply(marker_vals, function(vals_i) {
        if (is.matrix(vals_i))
            apply(vals_i, 2, mean)
        else 
            vals_i
    })
    names(marker_vals) <- rownames(pca_rotation)
    
    ## collapse the numeric vectors and create a matrix (the step will 
    ## remove the entries that have a list entry of NULL)
    marker_vals <- do.call("rbind", marker_vals) |>
        as.data.frame() |>
        tibble::rownames_to_column(var = "protein")
    tibble::tibble(marker_vals)
}

#' @name splitNames
#' 
#' @title Split the names when separated by ;
#' 
#' @description
#' The function \code{splitNames} creates a ";"-separated list of protein names. 
#' If \code{na.rm = TRUE}, the \code{NA} values within the protein names 
#' are removed. \code{NA} values might occur if protein names are translated
#' from one id to the other, e.g. from SYMBOL to Ensembl.
#' 
#' @details
#' Helper function for various functions in the \code{apoptosisQuantification}
#' package.
#' 
#' @param protein_names character
#' @param na.rm logical(1)
#' 
#' @examples
#' protein_names <- c("prot_1;prot_2;NA")
#' apoptosisQuantification:::splitNames(protein_names = protein_names,
#'     na.rm = FALSE)
#' apoptosisQuantification:::splitNames(protein_names = protein_names,
#'     na.rm = TRUE)
splitNames <- function(protein_names, na.rm = TRUE) {
    
    ## split the protein names
    protein_names <- strsplit(protein_names, split = ";")
    
    ## remove the NA values
    if (na.rm)
        protein_names <- lapply(protein_names, 
            function(prot_i) prot_i[prot_i != "NA"])
    
    ## return
    protein_names
}


#' @name combineLoadings
#' 
#' @title Combine the loadings/calculate the length of loading vectors
#' 
#' @description
#' The function \code{combineLoadings} calculates the length of loading vectors
#' across several principal components within the Cartesian coordinate system. 
#' 
#' The length of vectors is calculated along the principal coordinates
#' defined by \code{PC}, e.g. if \code{PC = c("PC1", "PC2")} the 
#' length is calculated along the loadings of principal components 1 and 2.  
#' 
#' @details 
#' The length of vectors is calculated using the classical way in the 
#' cartesian coordinate system. 
#' 
#' For 1D (e.g. length of vectors for PC1):
#' \deqn{
#'     \bar{\bar{a}} = \sqrt{a_1^2}
#' }
#' 
#' For 2D (e.g. length of vectors for PC1 and PC2): 
#' \deqn{
#'     \bar{\bar{a}} = \sqrt{a_1^2 + a_2^2}
#' }
#' 
#' For 3D (e.g. length of vectors for PC1, PC2, and PC3):
#' \deqn{
#'     \bar{\bar{a}} = \sqrt{a_1^2 + a_2^2 + a_3^2}
#' }
#' 
#' where $a_1$ is the loading of a data point for PC1, $a_2$ for PC2, $a_3$ for 
#' PC3, etc. 
#' 
#' @param loadings matrix or data.frame
#' @param PC character
#' 
#' @return numeric vector
#' 
#' @importFrom dplyr all_of as_tibble select pull
#' @importFrom MatrixQCvis transformAssay imputeAssay
#' @importFrom stats setNames
#' @importFrom SummarizedExperiment assay
#' @importFrom tibble is_tibble
#' 
#' @export
#' 
#' @examples 
#' library(MatrixQCvis)
#' library(SummarizedExperiment)
#' library(tibble)
#' 
#' f <- system.file("protein_datasets/tanzer2020.RDS", 
#'     package = "apoptosisQuantification")
#' se <- readRDS(f)
#' prot <- assay(se)
#' loadings <- prcomp(t(prot))$rotation |> 
#'     as.data.frame() |>
#'     rownames_to_column(var = "protein") |> 
#'     as_tibble()
#' combineLoadings(loadings = loadings, PC = c("PC1", "PC2")) 
combineLoadings <- function(loadings, PC = c("PC1", "PC2")) {
    
    if (!tibble::is_tibble(loadings))
        stop("'loadings' is not a tibble")
    
    ## select the PCs of interest
    names_loadings <- loadings |> 
        dplyr::pull("protein")
    loadings_pcs <- loadings |> 
        tibble::as_tibble() |>
        dplyr::select(dplyr::all_of(PC))
    
    ## square the loadings and calculate the sum
    loadings_pcs <- rowSums(loadings_pcs^2)
    
    ## to obtain the length, take the square root, return the values
    loadings_pcs <- sqrt(loadings_pcs) 
    stats::setNames(loadings_pcs, names_loadings)
}


#' @name createLoadingsTbl
#' 
#' @title Create a tibble containing the loading values for marker and 
#' non-marker proteins
#' 
#' @description
#' The function \code{createLoadingsTbl} creates a tibble containing the loading
#' values for marker and non-marker proteins. 
#' 
#' \code{createLoadingsTbl} returns a tibble with three columns:
#' \itemize{
#'     \item \code{protein}: name of protein (in case of multiple assignemt 
#'         separated by \code{";"}),
#'     \item \code{type}: either \code{"marker"} or \code{non-marker}
#' }
#'
#' @details
#' In case of multiple assignments of proteins, such a multiple protein is
#' treated as a marker protein if there is at least one marker protein
#' (defined in \code{names(vals_markers)}) in the multiple assignment.
#' 
#' @param vals_markers \code{numeric} vector
#' @param vals_all \code{numeric} vector
#' 
#' @return tibble
#' 
#' @export
#' 
#' @importFrom MatrixQCvis transformAssay imputeAssay
#' @importFrom SummarizedExperiment assay
#' @importFrom tibble tibble
#' @importFrom stats setNames
#' 
#' @examples 
#' library(MatrixQCvis)
#' library(SummarizedExperiment)
#'
#' ## proteomics example data
#' f <- system.file("protein_datasets/proteomics_example.RDS", 
#'     package = "apoptosisQuantification")
#' se <- readRDS(f)
#' prot <- assay(se) |> 
#'     transformAssay(method = "vsn") |> 
#'     imputeAssay(method = "MinDet")
#' markers <- readMarkers(type = "apoptosis")
#'     
#' ## get loadings of markers and of the complete data set
#' loadings_markers <- getLoadingsOfMarkers(prot = prot, markers = markers)
#' vals_all <- prcomp(t(prot))$rotation[, 1]
#' vals_markers <- setNames(loadings_markers$PC1, loadings_markers$protein)
#' createLoadingsTbl(vals_markers = vals_markers, vals_all = vals_all)
#' 
#' ## Tanzer et al. (2020)
#' ## source: https://www.ebi.ac.uk/pride/archive/projects/PXD014966
#' f <- system.file("protein_datasets/tanzer2020.RDS", 
#'     package = "apoptosisQuantification")
#' tanzer2020 <- readRDS(f)
#' 
#' ## get loadings of markers and of the complete data set
#' prot <- assay(tanzer2020)
#' markers <- readMarkers(type = "apoptosis")
#' loadings_markers <- getLoadingsOfMarkers(prot = prot, markers = markers)
#' vals_all <- prcomp(t(prot))$rotation[, 1]
#' vals_markers <- setNames(loadings_markers$PC1, loadings_markers$protein)
#' createLoadingsTbl(vals_markers = vals_markers, vals_all = vals_all)
createLoadingsTbl <- function(vals_markers = vals_markers, vals_all = vals_all) {
    
    if (!is.vector(vals_markers, mode = "numeric"))
        stop("'vals_markers' is not a numeric vector")
    if (!is.vector(vals_all, mode = "numeric"))
        stop("'vals_all' is not a numeric vector")
    
    ## split the marker proteins
    markers_proteins <- unlist(strsplit(names(vals_markers), split = ";"))

    all_proteins <- splitNames(protein_names = names(vals_all), na.rm = TRUE)
    ind_markers <- unlist(lapply(all_proteins, 
            function (proteins) any(proteins %in% markers_proteins)))

    ind_nonmarkers <- !ind_markers
    vals_nonmarkers <- vals_all[ind_nonmarkers]
    
    ## create data frames
    tbl_markers <- tibble::tibble(
        protein = names(vals_markers), type = "marker", value = vals_markers)
    tbl_nonmarkers <- tibble::tibble(
        protein = names(vals_all)[ind_nonmarkers], type = "non-marker", 
        value = vals_nonmarkers)
    
    ## combine the values and return
    tbl_comb <- rbind(tbl_markers, tbl_nonmarkers)
    tbl_comb
}

#' @name plotECDF
#' 
#' @title Plot empirical cumulative distribution function of loading values
#' 
#' @description 
#' The function \code{plotECDF} creates an empirical cumulative distribution
#' function (ECDF) plot of the loadings distribution of marker and non-markers.
#' 
#' @details 
#' The function \code{plotECDF} takes as input a \code{tibble} (\code{tbl}) 
#' that stores the loading values in the column \code{"value"} and the 
#' information on marker/non-marker in the column \code{"type"}.
#' 
#' @param tbl tibble
#' 
#' @return \code{gg}
#' 
#' @export
#' 
#' @importFrom ggplot2 ggplot aes_string stat_ecdf xlab ylab theme element_blank
#' @importFrom MatrixQCvis transformAssay imputeAssay
#' @importFrom stats setNames
#' @importFrom SummarizedExperiment assay
#' @importFrom tibble is_tibble
#' 
#' @examples 
#' library(MatrixQCvis)
#' library(SummarizedExperiment)
#' 
#' ## Tanzer et al. (2020)
#' ## source: https://www.ebi.ac.uk/pride/archive/projects/PXD014966
#' f <- system.file("protein_datasets/tanzer2020.RDS", 
#'     package = "apoptosisQuantification")
#' tanzer2020 <- readRDS(f)
#' 
#' ## get loadings of markers and of the complete data set
#' prot <- assay(tanzer2020)
#' markers <- readMarkers(type = "apoptosis")
#' loadings_markers <- getLoadingsOfMarkers(prot = prot, markers = markers)
#' vals_all <- prcomp(t(prot))$rotation[, 1]
#' vals_markers <- setNames(loadings_markers$PC1, loadings_markers$protein)
#' tbl <- createLoadingsTbl(vals_markers = vals_markers, vals_all = vals_all)
#' 
#' ## plot the distribution as ECDF
#' plotECDF(tbl)
plotECDF <- function(tbl) {
    
    ## check the object
    if (!tibble::is_tibble(tbl)) stop("'tbl' is not a tibble")
    cols <- colnames(tbl)
    if (!"protein" %in% cols) stop("'tbl' does not have column 'protein'")
    if (!"type" %in% cols) stop("'tbl' does not have column 'type'")
    if (!"value" %in% cols) stop("'tbl' does not have column 'value'")
    if (!is.numeric(tbl$value)) stop("'tbl$value' has to be numeric")
    
    ## do the actual plotting
    ggplot2::ggplot(data = tbl, 
        ggplot2::aes_string(x = "value", group = "type", col = "type")) +
        ggplot2::stat_ecdf(size = 1) + 
        ggplot2::xlab("loadings") + ggplot2::ylab("ECDF") + 
        ggplot2::theme(
            legend.title = ggplot2::element_blank(), legend.position = "top")
}

#' @name plotHistogram
#' 
#' @title Plot the histogram (distribution) of loading values 
#' 
#' @description 
#' The function \code{plotHistogram} creates (overlaid) histograms
#' plots of the loadings distribution of marker and non-markers.
#' 
#' @details 
#' The function \code{plotHistogram} takes as input a \code{tibble} (\code{tbl}) 
#' that stores the loading values in the column \code{"value"} and the 
#' information on marker/non-marker in the column \code{"type"}.
#' 
#' @param tbl tibble
#' 
#' @export
#' 
#' @return \code{gg}
#'
#' @importFrom dplyr mutate
#' @importFrom ggplot2 ggplot aes_string geom_histogram
#' @importFrom MatrixQCvis transformAssay imputeAssay
#' @importFrom stats setNames
#' @importFrom SummarizedExperiment assay
#' 
#' @examples 
#' library(MatrixQCvis)
#' library(SummarizedExperiment)
#' 
#' ## proteomics example data
#' f <- system.file("protein_datasets/proteomics_example.RDS", 
#'     package = "apoptosisQuantification")
#' se <- readRDS(f)
#' prot <- assay(se) |> 
#'     transformAssay(method = "vsn") |> 
#'     imputeAssay(method = "MinDet")
#' markers <- readMarkers(type = "apoptosis", fc = 2, n = 1)
#'     
#' ## get loadings of markers and of the complete data set
#' loadings_markers <- getLoadingsOfMarkers(prot = prot, markers = markers)
#' vals_all <- prcomp(t(prot))$rotation[, 1]
#' vals_markers <- setNames(loadings_markers$PC1, loadings_markers$protein)
#' tbl <- createLoadingsTbl(vals_markers = vals_markers, vals_all = vals_all)
#' plotHistogram(tbl)
#' 
#' ## Tanzer et al. (2020)
#' ## source: https://www.ebi.ac.uk/pride/archive/projects/PXD014966
#' f <- system.file("protein_datasets/tanzer2020.RDS",
#'     package = "apoptosisQuantification")
#' tanzer2020 <- readRDS(f)
#' 
#' ## get loadings of markers and of the complete data set
#' prot <- assay(tanzer2020)
#' markers <- readMarkers(type = "apoptosis", fc = 2, n = 1)
#' loadings_markers <- getLoadingsOfMarkers(prot = prot, markers = markers)
#' vals_all <- prcomp(t(prot))$rotation[, 2]
#' vals_markers <- setNames(loadings_markers$PC2, loadings_markers$protein)
#' tbl <- createLoadingsTbl(vals_markers = vals_markers, vals_all = vals_all)
#' 
#' plotHistogram(tbl)
plotHistogram <- function(tbl) {
    
    
    ## check the object
    if (!tibble::is_tibble(tbl)) stop("'tbl' is not a tibble")
    cols <- colnames(tbl)
    if (!"protein" %in% cols) stop("'tbl' does not have column 'protein'")
    if (!"type" %in% cols) stop("'tbl' does not have column 'type'")
    if (!"value" %in% cols) stop("'tbl' does not have column 'value'")
    if (!is.numeric(tbl$value)) stop("'tbl$value' has to be numeric")
    
    ## do the actual plotting
    tbl <- tbl |> 
        dplyr::mutate(value = abs(!!as.symbol("value")))
    ggplot2::ggplot(data = tbl, 
        ggplot2::aes_string(x = "value", col = "type", fill = "type")) +
        ggplot2::geom_histogram(alpha = 0.5, position = "identity")
}
    
#' @name performWilcoxonTest
#' 
#' @title Perform the Wilcoxon rank sum and signed rank tests on loading values
#' 
#' @description 
#' The function \code{performWilcoxonTest} performs a Wilcoxon rank sum and 
#' signed rank test on the loading values. The null hypothesis is that 
#' the mean of ranks of the marker loadings is smaller or equal to the 
#' ranks of the non-marker loadings. 
#' The null hypothesis can be rejected with probability \code{p}
#' (alternative hypothesis: loadings of markers are greater than the loadings
#' of non-markers). 
#' 
#' @details 
#' The function \code{performWilcoxonTest} takes as input a \code{tibble} 
#' (\code{tbl}) that stores the loading values in the column \code{"value"} 
#' and the information on marker/non-marker in the column \code{"type"}.
#' 
#' @param tbl tibble
#' 
#' @return htest
#' 
#' @export
#' 
#' @importFrom dplyr filter pull
#' @importFrom MatrixQCvis transformAssay imputeAssay
#' @importFrom stats wilcox.test setNames
#' @importFrom SummarizedExperiment assay
#' 
#' @examples 
#' library(MatrixQCvis)
#' library(SummarizedExperiment)
#' 
#' ## Tanzer et al. (2020)
#' ## source: https://www.ebi.ac.uk/pride/archive/projects/PXD014966
#' f <- system.file("protein_datasets/tanzer2020.RDS", 
#'     package = "apoptosisQuantification")
#' tanzer2020 <- readRDS(f)
#' 
#' ## get loadings of markers and of the complete data set
#' prot <- assay(tanzer2020)
#' markers <- readMarkers(type = "apoptosis")
#' loadings_markers <- getLoadingsOfMarkers(prot = prot, markers = markers)
#' vals_all <- prcomp(t(prot))$rotation[, 1]
#' vals_markers <- setNames(loadings_markers$PC1, loadings_markers$protein)
#' tbl <- createLoadingsTbl(vals_markers = vals_markers, vals_all = vals_all)
#' 
#' performWilcoxonTest(tbl)
performWilcoxonTest <- function(tbl) {
    
    ## check the object
    if (!tibble::is_tibble(tbl)) stop("'tbl' is not a tibble")
    cols <- colnames(tbl)
    if (!"protein" %in% cols) stop("'tbl' does not have column 'protein'")
    if (!"type" %in% cols) stop("'tbl' does not have column 'type'")
    if (!"value" %in% cols) stop("'tbl' does not have column 'value'")
    if (!is.numeric(tbl$value)) stop("'tbl$value' has to be numeric")
    
    ## do the actual test
    value_markers <- tbl |>
        dplyr::filter(!!as.symbol("type") == "marker") |>
        dplyr::pull("value")
    value_markers <- abs(value_markers)
    value_nonmarkers <- tbl |>
        dplyr::filter(!!as.symbol("type") == "non-marker") |> 
        dplyr::pull("value")
    value_nonmarkers <- abs(value_nonmarkers)

    ## perform the Wilcoxon test and return the test results
    ## alternative hypothesis is that the value of markers is greater than
    ## the value of non-markers
    wilcox.test(value_markers, value_nonmarkers, alternative = "greater")
}

